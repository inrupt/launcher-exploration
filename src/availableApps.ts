import { Reference } from 'tripledoc';
import { acl, schema } from 'rdf-namespaces';

export type Modes = typeof acl.Read | typeof acl.Append | typeof acl.Write | typeof acl.Control;
export interface ClassFileRequirement {
  forClass: Reference;
  requiredModes: Array<Modes>;
  public?: boolean;
  defaultFilename: string;
};
export interface PodWideRequirement {
  podWidePemissions: Array<Modes>;
}
export interface ContainerBoundrequirement {
  container: Reference;
  requiredModes: Array<Modes>;
}
export type Requirement = ClassFileRequirement | PodWideRequirement | ContainerBoundrequirement;

export function isClassFileRequirement(r: Requirement): r is ClassFileRequirement {
  return (typeof (r as ClassFileRequirement).forClass !== 'undefined');
}
export function isPodWideRequirement(r: Requirement): r is PodWideRequirement {
  return (typeof (r as PodWideRequirement).podWidePemissions !== 'undefined');
}
export function isContainerBoundRequirement(r: Requirement): r is ContainerBoundrequirement {
  return (typeof (r as ContainerBoundrequirement).container !== 'undefined');
}
/**
 * This can be used to have TypeScript warn you when another type is added to [[Requirement]]
 */
export function isExhaustive(_typeToCheck: never): never {
  throw new Error('Did not check for all possible values of a type.');
}

export interface Listing {
  launchUrl: string;
  name: string;
  tagline: string;
  screenshot?: string;
  requirements?: Requirement[];
};

export const availableApps: Listing[] = [
  {
    name: 'Notepod',
    tagline: 'A note-taking app for Solid',
    launchUrl: 'https://notepod.vincenttunru.com/',
    requirements: [
      {
        podWidePemissions: [],
      },
      {
        forClass: schema.TextDigitalDocument,
        requiredModes: [acl.Read, acl.Append, acl.Write],
        defaultFilename: 'notes',
        public: true,
      },
    ],
  },
  {
    name: 'Poddit',
    tagline: 'Private bookmarking',
    launchUrl: 'https://poddit.app',
    requirements: [
      {
        podWidePemissions: [],
      },
      {
        forClass: 'http://www.w3.org/2002/01/bookmark#Bookmark',
        requiredModes: [acl.Read, acl.Append, acl.Write],
        defaultFilename: 'bookmarks',
        public: true,
      },
    ],
  },
  {
    name: 'Generator demo',
    tagline: 'Output of inrupt\'s generator for Solid React Applications',
    launchUrl: 'https://generator.inrupt.com/',
    requirements: [
      {
        podWidePemissions: [acl.Read, acl.Append, acl.Write, acl.Control],
      },
      {
        container: 'tictactoe',
        requiredModes: [acl.Read, acl.Append, acl.Write, acl.Control],
      },
    ],
  },
  {
    name: 'Focus',
    tagline: 'Solid Task Manager',
    launchUrl: 'https://noeldemartin.github.io/solid-focus/',
    requirements: [
      {
        podWidePemissions: [acl.Read, acl.Append, acl.Write, acl.Control],
      },
    ],
  },
];
