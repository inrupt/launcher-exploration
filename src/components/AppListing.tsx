import React from 'react';
import { Listing } from '../availableApps';
import { Card, CardHeader, CardActions,  Typography, CardContent } from '@material-ui/core';
import { Screenshot } from './Screenshot';
import { LaunchButton } from './LaunchButton';
import { PrepareButton } from './PrepareButton';

interface Props {
  listing: Listing;
};

export const AppListing: React.FC<Props> = (props: Props) => {
  return (
    <Card>
      <Screenshot id={props.listing.screenshot}/>
      <CardHeader title={props.listing.name}/>
      <CardContent>
        <Typography variant="subtitle1">
          {props.listing.tagline}
        </Typography>
      </CardContent>
      <CardActions>
        <PrepareButton listing={props.listing}/>
        <LaunchButton listing={props.listing}/>
      </CardActions>
    </Card>
  );
};
