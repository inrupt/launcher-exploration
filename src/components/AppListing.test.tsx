import React from 'react';
import { render } from '@testing-library/react'
import { AppListing } from './AppListing';
import { Listing, Modes } from '../availableApps';
import { schema, acl } from 'rdf-namespaces';

jest.mock('@solid/react', () => ({
  LoggedOut: () => null,
  LoggedIn: 'div',
  useWebId: jest.fn().mockReturnValue('https://arbitrary-webid.com'),
}));

it('tells the user when no special permissions are required', () => {
  const listing: Listing = {
    launchUrl: 'https://example.com',
    name: 'Mock app listing',
    tagline: 'Mock tagline',
  };
  const { container } = render(<AppListing listing={listing}/>);
  expect(container.textContent).toMatch('This app does not require special permissions.');
});

describe('with permission requests for a specific class of data', () => {
  function getMockListing(requiredModes: Modes[]) {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          forClass: schema.Movie,
          defaultFilename: 'movies',
          requiredModes: requiredModes,
        }
      ],
    };
    return listing;
  }

  it('tells the user when only read permissions are required', () => {
    const listing = getMockListing([acl.Read]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only append permissions are required', () => {
    const listing = getMockListing([acl.Append]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only write permissions are required', () => {
    const listing = getMockListing([acl.Write]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only control permissions are required', () => {
    const listing = getMockListing([acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('renders (effectively equal) plain Write and Append+Write permissions in the same way', () => {
    const writeContainer = document.createElement('div');
    const writeListing = getMockListing([acl.Write]);
    const writeRender = render(<AppListing listing={writeListing}/>, { container: writeContainer });
    const writeRequirements = writeRender.getAllByRole('listitem');

    const appendAndWriteContainer = document.createElement('div');
    const appendAndWriteListing = getMockListing([acl.Append, acl.Write]);
    const appendAndWriteRender = render(<AppListing listing={appendAndWriteListing}/>, { container: appendAndWriteContainer });
    const appendAndWriteRequirements = appendAndWriteRender.getAllByRole('listitem');

    expect(writeRequirements).toEqual(appendAndWriteRequirements);
  });

  it('correctly lists all requested permissions when combined', () => {
    const listing = getMockListing([acl.Read, acl.Append, acl.Write, acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('shows a friendlier alias for `http://www.w3.org/2002/01/bookmark#Bookmark`', () => {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          forClass: 'http://www.w3.org/2002/01/bookmark#Bookmark',
          defaultFilename: 'bookmarks',
          requiredModes: [acl.Read],
        }
      ],
    };
    const { getByRole, getByTitle } = render(<AppListing listing={listing}/>);
    expect(getByRole('listitem').textContent).toBe('Read bookmarks in your Pod.');
    expect(getByTitle('http://www.w3.org/2002/01/bookmark#Bookmark').textContent).toBe('bookmarks');
  });

  it('shows a friendlier alias for `schema:TextDigitalDocument`', () => {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          forClass: schema.TextDigitalDocument,
          defaultFilename: 'bookmarks',
          requiredModes: [acl.Read],
        }
      ],
    };
    const { getByRole, getByTitle } = render(<AppListing listing={listing}/>);
    expect(getByRole('listitem').textContent).toBe('Read text files in your Pod.');
    expect(getByTitle(schema.TextDigitalDocument).textContent).toBe('text files');
  });
});

describe('with permission requests for a specific folder', () => {
  function getMockListing(requiredModes: Modes[]) {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          container: 'arbitrary-container',
          requiredModes: requiredModes,
        }
      ],
    };
    return listing;
  }

  it('tells the user when only read permissions are required', () => {
    const listing = getMockListing([acl.Read]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only append permissions are required', () => {
    const listing = getMockListing([acl.Append]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only write permissions are required', () => {
    const listing = getMockListing([acl.Write]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only control permissions are required', () => {
    const listing = getMockListing([acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('renders (effectively equal) plain Write and Append+Write permissions in the same way', () => {
    const writeContainer = document.createElement('div');
    const writeListing = getMockListing([acl.Write]);
    const writeRender = render(<AppListing listing={writeListing}/>, { container: writeContainer });
    const writeRequirements = writeRender.getAllByRole('listitem');

    const appendAndWriteContainer = document.createElement('div');
    const appendAndWriteListing = getMockListing([acl.Append, acl.Write]);
    const appendAndWriteRender = render(<AppListing listing={appendAndWriteListing}/>, { container: appendAndWriteContainer });
    const appendAndWriteRequirements = appendAndWriteRender.getAllByRole('listitem');

    expect(writeRequirements).toEqual(appendAndWriteRequirements);
  });

  it('correctly lists all requested permissions when combined', () => {
    const listing = getMockListing([acl.Read, acl.Append, acl.Write, acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });
});

describe('with permission requests for the entire Pod', () => {
  function getMockListing(requiredModes: Modes[]) {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          podWidePermissions: requiredModes,
        }
      ],
    };
    return listing;
  }

  it('tells the user when only read permissions are required', () => {
    const listing = getMockListing([acl.Read]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only append permissions are required', () => {
    const listing = getMockListing([acl.Append]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only write permissions are required', () => {
    const listing = getMockListing([acl.Write]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when only control permissions are required', () => {
    const listing = getMockListing([acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('renders (effectively equal) plain Write and Append+Write permissions in the same way', () => {
    const writeContainer = document.createElement('div');
    const writeListing = getMockListing([acl.Write]);
    const writeRender = render(<AppListing listing={writeListing}/>, { container: writeContainer });
    const writeRequirements = writeRender.getAllByRole('listitem');

    const appendAndWriteContainer = document.createElement('div');
    const appendAndWriteListing = getMockListing([acl.Append, acl.Write]);
    const appendAndWriteRender = render(<AppListing listing={appendAndWriteListing}/>, { container: appendAndWriteContainer });
    const appendAndWriteRequirements = appendAndWriteRender.getAllByRole('listitem');

    expect(writeRequirements).toEqual(appendAndWriteRequirements);
  });

  it('correctly lists all requested permissions when combined', () => {
    const listing = getMockListing([acl.Read, acl.Append, acl.Write, acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });
});

describe('with mixed types of permission requests', () => {
  function getMockListing(requiredPodWideModes: Modes[], requiredContainerModes: Modes[]) {
    const listing: Listing = {
      launchUrl: 'https://example.com',
      name: 'Mock app listing',
      tagline: 'Mock tagline',
      requirements: [
        {
          podWidePermissions: requiredPodWideModes,
        },
        {
          container: 'arbitrary-container',
          requiredModes: requiredContainerModes,
        }
      ],
    };
    return listing;
  }

  it('tells the user when only one mode for every type of permission is required', () => {
    const listing = getMockListing([acl.Read], [acl.Append]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });

  it('tells the user when multiple modes are required', () => {
    const listing = getMockListing([acl.Read], [acl.Read, acl.Write, acl.Control]);
    const { getAllByRole } = render(<AppListing listing={listing}/>);
    const requirements = getAllByRole('listitem');
    expect(requirements).toMatchSnapshot();
  });
});
