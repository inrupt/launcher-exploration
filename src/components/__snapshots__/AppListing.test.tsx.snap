// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with mixed types of permission requests tells the user when multiple modes are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete data from the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with mixed types of permission requests tells the user when only one mode for every type of permission is required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific class of data correctly lists all requested permissions when combined 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       from your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific class of data tells the user when only append permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific class of data tells the user when only control permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific class of data tells the user when only read permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific class of data tells the user when only write permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       from your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new 
      data of 
      <abbr
        title="http://schema.org/Movie"
      >
        a specific type
      </abbr>
       in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific folder correctly lists all requested permissions when combined 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete data from the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific folder tells the user when only append permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific folder tells the user when only control permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific folder tells the user when only read permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for a specific folder tells the user when only write permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete data from the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data in the folder 
      <code>
        arbitrary-container
      </code>
       on your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for the entire Pod correctly lists all requested permissions when combined 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to any data in your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete any data from your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data to your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for the entire Pod tells the user when only append permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data to your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for the entire Pod tells the user when only control permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Change who has access to any data in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for the entire Pod tells the user when only read permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Read all data in your Pod.
    </p>
  </li>,
]
`;

exports[`with permission requests for the entire Pod tells the user when only write permissions are required 1`] = `
Array [
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Delete any data from your Pod.
    </p>
  </li>,
  <li>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      Add new data to your Pod.
    </p>
  </li>,
]
`;
